presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -173,25 +173,37 @@ public synchronized void freeSpill(long bytes)
         spillSpaceTracker.free(bytes);
     }
 
-    public synchronized void setMemoryPool(MemoryPool pool)
+    public void setMemoryPool(MemoryPool pool)
     {
-        requireNonNull(pool, "pool is null");
-        if (memoryPool == pool) {
-            // Don't unblock our tasks and thrash the pools, if this is a no-op
-            return;
+        // this method acquires the monitors of both queryMemoryContext and
+        // this instance, and there can be a deadlock if they are not
+        // acquired in the same order.
+
+        // the thread calling this method should first acquire the lock
+        // on queryMemoryContext as that's the order driver threads
+        // acquire the monitors of these objects when they reserve memory
+        // through the child memory context -> ... -> root memory context -> query context
+        synchronized (queryMemoryContext) {
```

Comment:
Where exactly is monitor for `queryMemoryContext`taken in the codebase? I cannot find it. 

1. `MemoryTrackingContext` is not synchronized internally. It is marked as `@ThreadSafe` but I assume the fact is implication of its building blocks (just guessing - I do not know the code here). There is no explicit synchronization on `this`.  

2. Also I do not see any `synchronized(some instance of MemoryTrackingContext)` anywhere.

Reply:
@nezihyigitbasi  The `jstack` you shared suggest dead lock is caused by `RootAggregatedMemoryContext`, not `MemoryTrackingContext`.

Reply:
yes, updated.